Background images: add support for theme.json ref value resolution

ramonjd opened the pull request:
## What?

Part of:

- https://github.com/WordPress/gutenberg/issues/54336

So, you know how you can "reference" another value in theme.json using `"ref"`?

Well,  this PR adds support for ref resolution to `"background"` style properties.


## Why?

The `"background"` style properties haven't supported this feature, until now!

The theme.json schema has already been updated to reflect this change.

## How?

By passing the resolved background image value to the style engine.

In the client (editor JS), I'm doing it long-hand, as you can see. It's a bit fugly. In both the frontend (PHP) and editor, the style engine does the work of escaping `background-image` values and also ensuring they're wrapped in `url()`. 

The `WP_Theme_JSON` class resolves values before passing to the style engine, however the editor does it the other way around. 

However, the style engine does not perform any "ref" resolution. And since the work of escaping and wrapping in `url()` needs to happen after ref resolution, we need to make an exception for background image.


## Alternatives

1. We add logic to `WP_Theme_JSON` and `use-global-styles-output.js` to sanitize/build `url()` values and bypass the style engine for `background-image` values. Involves duplicating code.

3. The style engine could handle ref resolution (and possibly defaults)  via a public API, used both internally and externally. Theme.json tree and defaults could be passed as options. 

Example (not prescriptive):

```js
// Input
const themeJson = {
   styles: {
    background: {
         backgroundImage: {
             url: 'http://some.host/someimage.gif',
         }
     }
   }
};

const blockStyles = {
    background: {
         backgroundImage: { ref: 'styles.background.backgroundImage' }
    }
};

// Output
const styles = getCSSRules( blockStyles, {
     themeJson,
} );

// Somewhere in getCSSRules

const resolvedRefValue = get( options.themeJson, blockStyles.background.backgroundImage.ref );
const backgroundImageValue = ``url( '${ encodeURI( safeDecodeURI( resolvedRefValue.url ) ) }' )`
```


## Testing Instructions

1. Create a theme.json with valid background ref values (see theme.json example below). The example below also includes a relative path to an image.
2. Ensure that the values resolve in the frontend and site editor (and post editor for the heck of it)
3. Check that, when you replace a block image in global styles, that the image appears correctly in the frontend and backend.

<details>

<summary>Example block HTML</summary>

```html

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group">
<!-- wp:paragraph -->
<p>Group</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:verse -->
<pre class="wp-block-verse">Verse</pre>
<!-- /wp:verse -->

<!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:paragraph -->
<p>Quote</p>
<!-- /wp:paragraph --></blockquote>
<!-- /wp:quote -->

```

</details>

<details>

<summary>Example theme.json</summary>

```json


{
	"$schema": "../../schemas/json/theme.json",
	"version": 3,
	"settings": {
		"appearanceTools": true,
		"layout": {
			"contentSize": "840px",
			"wideSize": "1100px"
		}
	},
	"styles": {

		"background": {
			"backgroundImage": {
				"url": "https://images.pexels.com/photos/105819/pexels-photo-105819.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr="
			},
			"backgroundSize": "cover",
			"backgroundAttachment": "fixed",
			"backgroundPosition": "top left"
		},
		"blocks": {
			"core/verse": {
				"background": {
					"backgroundImage": {
						"url": "file:./img/1024x768_emptytheme_test_image.jpg"
					},
					"backgroundAttachment": "fixed",
					"backgroundSize": "61px"
				},
				"color": {
					"text": "white"
				},
				"dimensions": {
					"minHeight": "100px"
				}
			},
			"core/quote": {
				"background": {
					"backgroundImage": {
						"ref": "styles.background.backgroundImage"
					}
				},
				"dimensions": {
					"minHeight": "100px"
				}
			},
			"core/group": {
				"background": {
					"backgroundImage": {
						"ref": "styles.blocks.core/verse.background.backgroundImage"
					},
					"backgroundAttachment": {
						"ref": "styles.blocks.core/verse.background.backgroundAttachment"
					},
					"backgroundSize": {
						"ref": "styles.blocks.core/verse.background.backgroundSize"
					},
					"backgroundPosition":  {
						"ref": "styles.background.backgroundPosition"
					}
				},
				"color": {
					"text": {
						"ref": "styles.blocks.core/verse.color.text"
					}
				},
				"dimensions": {
					"minHeight": "111px"
				}
			}
		}
	}
}





```

</details>

<img width="901" alt="Screenshot 2024-07-31 at 3 37 45 PM" src="https://github.com/user-attachments/assets/b1a9fa3a-9b5a-4b4c-ad56-8bbc6baf1402">

## TODO

Split this PR up into (in order of merge priority):

- [x] Abstracting the ref and theme URI resolution in the client and tests https://github.com/WordPress/gutenberg/pull/64182
- [x] Ensuring only user-added images get default block styles https://github.com/WordPress/gutenberg/pull/64192
- [x] Adding support for background theme.json ref resolution (this PR)
- [ ] Export `getCSSVarFromStyleValue` from style engine in place of `compileStyleValue` (new PR) https://github.com/WordPress/gutenberg/pull/64490

ramonjd committed:
- resolve refs before passing styles to style engine in wp_theme_json
- add tests

- adds tests

Check for string types in background image so it will apply defaults for blocks.
Update comments

ramonjd commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2309,7 +2309,7 @@ protected static function flatten_tree( $tree, $prefix = '', $token = '--' ) {
 	 * array(
 	 *   'name'  => 'property_name',
-	 *   'value' => 'property_value,
+	 *   'value' => 'property_value',
```
Just fixing a small typo

ramonjd commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2361,10 +2362,13 @@ protected static function compute_style_properties( $styles, $settings = array()
-				$background_styles = gutenberg_style_engine_get_styles( array( 'background' => $styles['background'] ) );
-				$value             = $background_styles['declarations'][ $css_property ] ?? $value;
+			// Processes background image styles.
+			if ( 'background-image' === $css_property && ! empty( $value ) ) {
```
We never really needed to pass all global styles background styles to the style engine because, aside from `backgroundImage`, none of them needed any processing. 

The style engine sanitizes the `backgroundImage`  image URI and wraps it in `url()`.

`$value` already contains resolved refs thanks to [$value = static::get_property_value](https://github.com/WordPress/gutenberg/pull/64128/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7R2353) above.

ramonjd replied:
This needs a bit of refactor now the following PR has been merged:

- https://github.com/WordPress/gutenberg/pull/64192

The defaults need to be applied. 

Something like:

```php
if ( static::ROOT_BLOCK_SELECTOR !== $selector && ! empty( $styles['background']['backgroundImage']['id'] ) ) {
	if ( 'background-size' === $css_property && empty( $value ) ) {
		$value = 'cover';
	}
	// If the background size is set to `contain` and no position is set, set the position to `center`.
	if ( 'background-position' === $css_property && empty( $value ) && 'contain' === $styles['background']['backgroundSize'] ) {
		$value = 'center';
	}
}
```

ramonjd replied:
Probably in a new PR so it's easier to test.

ramonjd commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2452,11 +2456,14 @@ protected static function get_property_value( $styles, $path, $theme_json = null
-			$ref_value  = _wp_array_get( $theme_json, $value_path );
+			$ref_value  = _wp_array_get( $theme_json, $value_path, null );
+			// Background Image refs can refer to a string or an array containing a URL string.
+			$ref_value_url = $ref_value['url'] ?? null;
```
This is specifically for background images. 

It's part of the reason why I think this could be a style engine util, used both internally, and externally by `WP_Theme_JSON`.

The style engine could subsume many style-related utils and potentially reduce the site of `WP_Theme_JSON`.

aaronrobertshaw replied:
Anything that can chip away at the size of `WP_Theme_JSON` sounds like a win to me.

ramonjd commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -395,17 +395,40 @@ export function getStylesDeclarations(
+		 * 2. Resolve dynamic values before they are compiled by the style engine,
+		 * which doesn't (yet) resolve dynamic values.
+		 */
+		if ( blockStyles.background?.backgroundImage?.ref ) {
```
A great deal of custom code just for background images.

It works, but could be reduce by moving ref resolution into the style engine. The condition is that we'll need to pass a theme.json object to the style engine methods.

ramonjd commented on packages/style-engine/src/styles/background/index.ts:
```diff
@@ -21,20 +27,12 @@ const backgroundImage = {
-		}
-
-		return [];
+		return generateRule(
```
This will return the default value instead of `[]`. 

Optional, but it makes it consistent with other style engine methods.

The consequence is that, if the input value cannot be processed, it'll be returned and not wiped from the face of the app.

andrewserong commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -395,17 +395,40 @@ export function getStylesDeclarations(
+	 * 1. Sets default values for blocks (not root).
+	 * 2. .
+	 *
+	 * Note: A refactor is for the style engine to handle ref resolution (and possibly defaults)
+	 * via a public util used internally and externally. Theme.json tree and defaults could be passed
+	 * as options.
```
Tiny nit: there's a comment below that's fairly similar to this:

```
// The goal is to move everything to server side generated engine styles
// This is temporary as we absorb more and more styles into the engine.
```

Can we consolidate the two comments?

ramonjd replied:
Oh yeah, those comments 😄 

I'm not sure what the goal is anymore and whether moving everything to server is side actually feasible in the site editor, whose CSS needs to be generated and regenerated on the fly.

I'll fix it up.

andrewserong commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -395,17 +395,40 @@ export function getStylesDeclarations(
-	 * Applies to all background styles except the top-level site background.
+	 * Preprocess background image values.
+	 * 1. Sets default values for blocks (not root).
+	 * 2. .
```
Should this be something like the following?

```suggestion
	 * 2. Resolve dynamic values before they are compiled by the style engine.
```

andrewserong reviewed (commented):
Nice work! I've only taken it for a quick spin so far, but can give this some more testing tomorrow. It's mostly testing well for me 👍 

I noticed that it looks like the default of `cover` background size, and the thumbnail for the image in the UI control don't appear to be included when using the `ref`-based value. Are you able to replicate?

<img width="1078" alt="image" src="https://github.com/user-attachments/assets/66db1768-277a-4a60-acb0-e0b87a90c582">

Tiny bit of `theme.json` snippet:

```
			"core/quote": {
				"background": {
					"backgroundImage": {
						"ref": "styles.background.backgroundImage"
					}
				},
```

Overall I like your idea of abstracting things away in the style engine in the future. For now, though, it doesn't look like this has added too much code so far, so seems a reasonable approach to me! But curious to hear what other folks think, too 🙂

andrewserong commented:
I think I just noticed another issue: I wasn't able to update a block in global styles with a user set image. The `ref` based one set in `theme.json` appears to stick. Here's what it looks like for a Quote block when giving it a background image at the block level in global styles (note the thumbnail has the "correct" image uploaded by the user, but the editor canvas still shows the one derived from the `ref`):

<img width="1046" alt="image" src="https://github.com/user-attachments/assets/bed883e3-eb8c-48b7-9dc5-f789f1939d5f">

ramonjd commented:
Thanks for testing @andrewserong 

I'll give it belt tomorrow and straighten things out.

ramonjd commented:
> I noticed that it looks like the default of cover background size, and the thumbnail for the image in the UI control don't appear to be included when using the ref-based value.

Very interesting. So I think we never output defaults at the block level in theme.json - I'll have to fix that in general.

As for the thumbnail, we're passing `inheritedValues` to the block panel controls, but we're not resolving the refs. 

All fixable!

Thanks again for testing 🙇🏻

andrewserong commented:
> Very interesting. So I think we never output defaults at the block level in theme.json - I'll have to fix that in general.

Oh, good point! I can't remember the discussion now... maybe we decided that we'd only set the default when a user uploads versus when someone is using `theme.json` since in `theme.json` folks can be explicit about what they want to include? Happy to take your lead on what you think makes sense here 🙂

ramonjd commented:
> maybe we decided that we'd only set the default when a user uploads versus when someone is using theme.json since in theme.json folks can be explicit about what they want to include?

I _think_ that was the intention, looking at the [commit](https://github.com/WordPress/gutenberg/pull/60100/commits/7988ba533e34cb8cdbd980c604c44d0e97764030) at the time. 

I can't find any specific conversation though.

Adding defaults only to user-uploaded backgrounds makes sense to me - I'll test it through.

ramonjd committed:
- consolidate all resolutions in the background panel (image and ref)
- remove useGlobalStyleLinks completely
- delete theme URI utils

Remove string condition

LInto

ramonjd commented on packages/block-editor/src/components/global-styles/background-panel.js:
```diff
@@ -334,6 +340,7 @@ function BackgroundImageControls( {
 					! positionValue && ( 'auto' === sizeValue || ! sizeValue )
 						? '50% 0'
 						: positionValue,
+				backgroundSize: sizeValue,
```
Uploaded images will receive any incoming default values. 

For example, `hooks/background.js` defines `"cover"` - this will affect blocks only.

ramonjd commented on packages/block-editor/src/components/global-styles/background-panel.js:
```diff
@@ -679,6 +682,54 @@ export default function BackgroundPanel( {
+			}
+		);
+		return resolvedValues;
+	}, [ globalStyles, inheritedValue ] );
```
Here I'm consolidating all the data and resolution logic. 

That now includes `themeURIs` for relative URIs, which really only affects background images.

It's a first step towards linking globalStyles with all styles controls, at least for one panel. 

Where themeURIs have been passed around as props before, I've removed it.

ramonjd commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -54,21 +56,6 @@ const BLOCK_SUPPORT_FEATURE_LEVEL_SELECTORS = {
 };
 const { kebabCase } = unlock( componentsPrivateApis );
 
-function compileStyleValue( uncompiledValue ) {
```
Moved to utils.

ramonjd commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -424,17 +432,10 @@ export function getStylesDeclarations(
-			if ( ! ruleValue || ruleValue?.ref ) {
-				return;
-			}
+		let ruleValue = getResolvedRefValue( rule.value, tree, undefined );
```
I wonder if we could also resolve themeURIs here to removing the necessity to do it in [useGlobalStylesOutputWithConfig](https://github.com/WordPress/gutenberg/blob/43e9107e4ea6cb0115ed08af2ce794d2867f7246/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L1402-L1401)

I don't know, I'll test.

ramonjd commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -63,7 +60,7 @@ export function setBackgroundStyleDefaults( backgroundStyle ) {
 
 	// Set block background defaults.
-	if ( !! backgroundImage?.url ) {
+	if ( !! backgroundImage?.url || typeof backgroundImage === 'string' ) {
```
Note to self, check if this is still required. 

It's supposed to include `backgroundImage: url('')` values.

ramonjd replied:
I don't think so as `backgroundImage` with string values will inevitably come from theme.json. 

Given the logic in https://github.com/WordPress/gutenberg/pull/64192, default values shouldn't be applied to those values.

ramonjd reviewed (commented):
Another round of updates to address:

- resolving ref-based image values for thumbnails
- ensuring only uploaded background images in block get the block defaults, e.g., "cover"

ramonjd commented:
This PR does too much:

- resolves refs for background images
- centralizes clientside code for ref and URI path resolution and does refactoring around global styles
- adds tests

I'll split it up  (in order of merge priority):

-  Abstracting the ref and theme URI resolution in the client and tests (new PR)
-  Adding support for background theme.json ref resolution (this PR)
-  Export getCSSVarFromStyleValue from style engine in place of compileStyleValue (new PR)

andrewserong commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -395,21 +381,40 @@ export function getStylesDeclarations(
+		 * Set default values for block background styles for uploaded images.
+		 * Top-level styles are an exception as they are applied to the body.
+		 */
+		if ( ! isRoot && !! blockStyles.background?.backgroundImage?.id ) {
```
Not necessarily something to worry about just now, but what if we allowed external urls at some point via the UI? I suppose in that case, we could set the defaults when the user adds the external url, rather than worrying about it at this part of the process. So I think that means this code is probably fine!

ramonjd replied:
That's a good 	question. Yeah, I like your idea of setting some defaults here. I'm sure we'll find a way to identify what's been added by the user in the editor. 

I think 🙃

andrewserong replied:
Yeah, totally. Again, not a blocker for now!

andrewserong commented on packages/edit-site/src/components/global-styles/background-panel.js:
```diff
@@ -16,7 +16,6 @@ const BACKGROUND_DEFAULT_VALUES = {
 const {
 	useGlobalStyle,
 	useGlobalSetting,
-	useGlobalStyleLinks,
```
I like how the change in this PR means this part is happening internally to the panel now 👍

andrewserong commented:
> I'll split it up (in order of merge priority):

Sounds great, I like the direction of the refactors in this PR so far!

> Another round of updates to address:
> resolving ref-based image values for thumbnails

Has that been addressed? When re-testing, it looks like I'm still seeing an "Add background image" button at the block level in global styles, and in the block inspector when selecting a block instance. Just wanted to double-check in case I've messed something up in my testing locally 🙂

<img width="897" alt="image" src="https://github.com/user-attachments/assets/e7dfc5de-8957-47cc-b24c-2d118a8ffa44">

andrewserong commented:
Update: in my screenshot above, my root `styles.background.backgroundImage` was a string that used an external url. When I update to using a theme url, it works fine:

<img width="911" alt="image" src="https://github.com/user-attachments/assets/abfe5acf-5286-456e-8eb0-4834c575d17f">

ramonjd commented:
> in my screenshot above, my root styles.background.backgroundImage was a string that used an external url.

Ah, that would be it 👍🏻 

That was a concession we made a few cycles ago, that is, not to try to prioritize parsing out the image URI from such string values because those values could be anything, not just `url('happy-path.jpg')`.

I suppose we could do a simple regex and URI sanitization in a follow up. I'll add it to the tracking issue as an enhancement.

andrewserong commented:
> I suppose we could do a simple regex and URI sanitization in a follow up. I'll add it to the tracking issue as an enhancement.

Yeah, not a blocker at all! I wouldn't worry about it for now, just glad we got to the bottom of it 😄

ramonjd committed:
Reconcile global styles default value assignment an ref resolution after rebase.

ramonjd committed:
Update tests and perform check for background size before use.

ramonjd committed:
Fixes a but where a ref value will be resolved even where a url value exists. This will occur when base and user configs are merged.

Update tests and perform check for background size before use.

Update tests and perform check for background size before use.

ramonjd committed:
The commit tries to ensure that backgroundImage style objects are not merged, but replaced when merged theme.json configs. So, for example, an incoming config such as a saved user config should overwrite the background styles of the base theme.json, where it exists.

ramonjd commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3247,6 +3258,27 @@ public function merge( $incoming ) {
+			 * Background image styles should be replaced, not merged,
+			 * as they themselves are specific object definitions for the style.
+			 */
+			$background_image_path = array_merge( $path, static::PROPERTIES_METADATA['background-image'] );
```
@aaronrobertshaw I need your smarts here. 

On the frontend end, the objects are being replaced fine _mostly_.

There's a final merge coming from `gutenberg_enqueue_global_styles` for which this line borks the operation:

https://github.com/WordPress/gutenberg/blob/5a399c1ae2c50b7a7a6daaf06c6dae3f4a264e30/lib/script-loader.php#L43-L43

I ran out of time to dig deeper, but commenting out that filter makes things work.

I noticed you removed that filter in some [block variation operations](https://github.com/WordPress/gutenberg/blob/5a399c1ae2c50b7a7a6daaf06c6dae3f4a264e30/lib/block-supports/block-style-variations.php#L163-L163).

aaronrobertshaw replied:
I'll take a look 👍 

From memory, removing those filters was to ensure the relevant block type data was still within the overall theme json data set. That way the block style variations support could retrieve the appropriate variation style values regardless of whether block assets were loading separately or not.

aaronrobertshaw replied:
At a quick glance, if we're expecting to work with the full unfiltered set of style nodes here, we might need to toggle that filter on and off like the block style variations support.

Another thing to double-check will be if there's a Gutenberg and Core version of that filter. Both would need handling.

ramonjd replied:
Thanks for looking. 

I forgot to add, the way to test is:

- using the test theme.json in the PR description,
- add a custom image for the quote block in global styles. 
- check the frontend. 
- the core/quote block should display your custom image, but it displays the one defined in the theme.json ref
- the reason is, as far as I can tell, it's not running through the new `merge()` changes

It could be something buggy in the code - I'll keep testing too.

ramonjd replied:
I think I might have bypassed it in [38b0e1d](https://github.com/WordPress/gutenberg/pull/64128/commits/38b0e1d8473c84411eb82072f21cfd63283ba910) by using `get_styles_block_nodes`, which doesn't apply the filter.

aaronrobertshaw replied:
Nice option. I clearly wasn't appreciating that we were working internally to `WP_Theme_JSON` here 🤦

aaronrobertshaw replied:
After following the test instructions above, this looks to be working well for me.

ramonjd committed:
Fix bug where the style node list was being filtered, add comment.

aaronrobertshaw commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2381,21 +2382,28 @@ protected static function compute_style_properties( $styles, $settings = array()
+			/*
+			 * Processes background image styles.
+			 * If the value is a URL, it will be converted to a CSS `url()` value.
+			 * For uploaded image (images with a database ID), apply size and position defaults,
```
```suggestion
			 * For an uploaded image (images with a database ID), apply size and position defaults
```


Nit: Tweak to improve wording

aaronrobertshaw replied:
Updated in https://github.com/WordPress/gutenberg/pull/64128/commits/aaf0cc96e07c20dcbc186023efe12d0d0bcfcbe4

aaronrobertshaw commented on packages/block-editor/src/components/global-styles/utils.js:
```diff
@@ -589,6 +589,11 @@ export function getResolvedRefValue( ruleValue, tree ) {
 
+	/*
+	 * This converts references to a path to the value at that path
+	 * where the value is an array with a "ref" key, pointing to a path.
```
```suggestion
	 * Where the rule value is an object with a 'ref' property pointing
	 * to a path, this converts that path into the value at that path.
```

I know this comment was taken from the existing one in the PHP class but it reads poorly to me. Could we improve it? Maybe something along the lines of the suggestion above?

aaronrobertshaw replied:
I tweaked the wording in [936f216](https://github.com/WordPress/gutenberg/pull/64128/commits/936f2163e9cb64ddd8cde57f135e4d8f37140d65). Feel free to revert or improve as desired.

aaronrobertshaw commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2295,6 +2295,110 @@ public function test_merge_incoming_data_presets_use_default_names() {
+					),
+					'core/verse' => array(
+						'background' => array(
+							'backgroundImage' => "url(''http://example.org/verse.png')",
```
```suggestion
							'backgroundImage' => "url('http://example.org/verse.png')",
```

Typo

aaronrobertshaw replied:
Fixed via https://github.com/WordPress/gutenberg/pull/64128/commits/00ef141c181be613fabd486b4f59721597b982da

aaronrobertshaw commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2295,6 +2295,110 @@ public function test_merge_incoming_data_presets_use_default_names() {
+							'backgroundAttachment' => 'fixed',
+						),
+					),
+					'core/quote' => array(
```
Should these quote block styles also be in the expected results?

aaronrobertshaw replied:
Ah, I see the issue, the block styles here aren't under `styles.blocks`. I'll fix that up once I finished code review and some manual testing.

aaronrobertshaw replied:
Fixed via https://github.com/WordPress/gutenberg/pull/64128/commits/00ef141c181be613fabd486b4f59721597b982da

aaronrobertshaw commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -4916,6 +5020,60 @@ public function test_get_block_background_image_styles() {
+	/**
+	 * Testing background dynamic properties in theme.json.
+	 */
+	public function test_get_resolved_background_image_styles() {
```
For this test could we use `ReflectionClass` to test `get_property_value` rather than generating the full CSS? Or is this test aiming to also cover default values for size, position, etc?

ramonjd replied:
Good question. 

I was wondering whether `get_property_value` deserves testing on its own - I don't think it would hurt, since it fits the definition of a "unit" in testing. 

To answer your question, yeah, I wanted to also cover the other properties since none of them are resolved in trunk at the moment.

aaronrobertshaw replied:
The current approach covers all the desired bases and is probably less bloated than creating a data provider to run through all the different properties to ensure the correct resolution of ref values. 

Leaving it as it is, works for me 👍

aaronrobertshaw reviewed (commented):
Thanks for the explanatory comments and detailed PR description @ramonjd 👍 

I've followed the test instructions, including those in https://github.com/WordPress/gutenberg/pull/64128#discussion_r1714759106, and it is working pretty well for me.

I think I'm missing some context around then removal of the `_links` data and theme file uri stuff. I think it makes sense. Was this part of what you were planning to split into separate PRs?

On the code side of things, I noticed a few minor issues that I'll push commits for shortly. Feel free to revert anything you disagree with 🙂 

Once the dust settles, I'll give it another proper test tomorrow.

aaronrobertshaw committed:
Fix unit test for merging background image styles

aaronrobertshaw committed:
Tweak wording for readability

aaronrobertshaw committed:
Tweak comment

ramonjd commented:
Huge thanks for testing this one and pushing those changes @aaronrobertshaw LGTM

> I think I'm missing some context around then removal of the _links data and theme file uri stuff. I think it makes sense. 

I was pushing the _links data around from component to component, when I really only needed in the background-panel component.

The _links data is used to resolve any relative paths.

Most of the ground work for this was already done in https://github.com/WordPress/gutenberg/pull/64182, so this PR is mainly putting all the resolution operations required for background panels in one place:

https://github.com/WordPress/gutenberg/pull/64128/files#diff-94e57b126656d423c5f8bf48b9a23351cc4f562fa47bace9a4dd1e9f0150a522R727



> Was this part of what you were planning to split into separate PRs?

It's kinda related. My idea was to abstract what the [background panel is doing in terms of resolution](https://github.com/WordPress/gutenberg/pull/64128/files#diff-94e57b126656d423c5f8bf48b9a23351cc4f562fa47bace9a4dd1e9f0150a522R716-R737) so, eventually other styles could take advantage. 

It was a bit of forward work for the following issue:

- https://github.com/WordPress/gutenberg/issues/37752

So, not really required right now, and I think it makes more sense to do afterwards.


I do see a couple of follow ups to this PR for the very immediate or for the longer term scope. In opinionated order of least complex to most complex:


- [ ] Export `getCSSVarFromStyleValue` from style engine in place of `compileStyleValue` (new PR)
- [ ] Generally, move some style utils to the style engine:
    - E.g., resolution of ref and other values, see comment https://github.com/WordPress/gutenberg/pull/64128/files#r1697931036
    - export `getCSSVarFromStyleValue` from style engine in place of `compileStyleValue` (new PR)
- [ ] As mentioned above, abstract all ref/URI resolving to a separate hook or whatever so any block style control can use it.

ramonjd committed:
Reduce the scope of the merge PHP unit test - let's just test background styles as general styles are tested above in test_merge_incoming_data

aaronrobertshaw reviewed (approved):
Nice work wrangling all this @ramonjd ✨ 

Special thanks for the detailed PR description and extra explanations to help get me up to speed here.

This tests well for me across the board.

✅ Theme.json set background styles apply correctly in both editors and frontend
✅ Background ref values resolve correctly
✅ Updating Global Styles overrides background image values, including ref values
✅ Updating Global Styles background image that is the source for another block's ref value does get reflected on the block referencing it
✅ Global Styles changes apply consistently across editors and frontend
✅ Block instance styles override Global Styles correctly

LGTM 🚢

kevin940726 reviewed (commented):
Thanks for the PR!

I tested it but noticed something I can't explain. I'm not sure if there's anything I'm doing wrong or this is not even related to this PR?

When I set the global styles background for `core/quote` to a custom image like shown below, the quote block is still referencing the background image.

![image](https://github.com/user-attachments/assets/a563d4d2-f559-49b0-b1d0-cdb7b3112630)

However, the frontend is correctly showing the custom image I uploaded.

![image](https://github.com/user-attachments/assets/212f46c6-ba8d-4ef4-9457-18bbb594ad51)

The block itself doesn't have per-block style.

You can also find the inconsistency on the group block above 🤔.  I'm not too familiar with the feature and the code base so I could use some sanity check here 🙇.

ramonjd commented:
Thanks for testing @kevin940726 and @aaronrobertshaw 🙇🏻 

> When I set the global styles background for core/quote to a custom image like shown below, the quote block is still referencing the background image.

I can't seem to reproduce 🤔 I checked in Firefox, Chrome and Safari.


https://github.com/user-attachments/assets/458e9768-6748-4f16-b823-e63e7c046980

I recognize the behavior, as it was the reason behind [this change](https://github.com/WordPress/gutenberg/pull/64128/files#diff-70e38d517891def9ad3dab3101260b1939bd4f6f0a60eb067ebbd33f6752cc50R37).

When uploading an image there's a slight delay I see, but that's because we're waiting on the upload callback to fire.

Which makes me wonder if we should have some loading UI there to indicate that an image is being uploaded, like the image block has 🤔 

I'll add it as a follow up 👍🏻

aaronrobertshaw commented:
I'm also having trouble replicating this issue (https://github.com/WordPress/gutenberg/pull/64128#pullrequestreview-2238656011).

The global styles background image for the quote block gets applied correctly for me. I tried a few approaches and all worked well:

- Selecting an existing image from the media library
- Uploading a new image
- Throttling the connection to slow 3g and uploading a new image

A loading UI as a follow-up sounds like a nice touch to me 👍 

<img width="1512" alt="Screenshot 2024-08-15 at 10 00 56 AM" src="https://github.com/user-attachments/assets/0e2695b3-8cfd-43b4-a319-10d1d1d04049">

kevin940726 reviewed (approved):
Well, I certainly can't reproduce https://github.com/WordPress/gutenberg/pull/64128#pullrequestreview-2238656011 either 🤦. I'm guessing an outdated or branch locally on my end. Sorry for the false alarm!

In other words, this works well in my testing! Thanks!
